keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,625 @@
+         * [Changes Required](#changes-required)
+
+## Summary
+
+This proposal adds IPv4/IPv6 dual stack functionality to Kubernetes clusters. This includes the following concepts:
+- Awareness of multiple IPv4/IPv6 address assignments per pod
+- Native IPv4-to-IPv4 in parallel with IPv6-to-IPv6 communications to, from, and within a cluster
+
+## Motivation
+
+The adoption of IPv6 has increased in recent years, and customers are requesting IPv6 support in Kubernetes clusters. To this end, the support of IPv6-only clusters was added as an alpha feature in Kubernetes Version 1.9. Clusters can now be run in either IPv4-only, IPv6-only, or in a "single-pod-IP-aware" dual-stack configuration. This "single-pod-IP-aware" dual-stack support is limited by the following restrictions:
+- Some CNI network plugins are capable of assigning dual-stack addresses on a pod, but Kubernetes is aware of only one address per pod.
+- Kubernetes system pods (api server, controller manager, etc.) can have only one IP address per pod, and system pod addresses are either all IPv4 or all IPv6.
+- Endpoints for services are either all IPv4 or all IPv6 within a cluster.
+- Service IPs are either all IPv4 or all IPv6 within a cluster.
+
+For scenarios that require legacy IPv4-only clients or services (either internal or external to the cluster), the above restrictions mean that complex and expensive IPv4/IPv6 transition mechanisms (e.g. NAT64/DNS64, stateless NAT46, or SIIT/MAP) will need to be implemented in the data center networking.
+
+One alternative to adding transition mechanisms would be to modify Kubernetes to provide support for IPv4 and IPv6 communications in parallel, for both pods and services, throughout the cluster (a.k.a. "full" dual stack).
+
+A second, simpler alternative, which is a variation to the "full" dual stack model, would be to provide dual stack addresses for pods and nodes, but restrict service IPs to be single-family (i.e. allocated from a single service CIDR). In this case, service IPs in a cluster would be either all IPv4 or all IPv6, as they are now. Compared to a full dual-stack approach, this "dual-stack pods / single-family services" approach saves on implementation complexity, but would introduce some minor feature restrictions. (For more details on these tradeoffs, please refer to the "Variation: Dual-Stack Service CIDRs" section under "Alternatives" below).
+
+This proposal aims to add "dual-stack pods / single-family services" support to Kubernetes clusters, providing native IPv4-to-IPv4 communication and native IPv6-to-IPv6 communication to, from and within a Kubernetes cluster.
+
+### Goals
+
+- Pod Connectivity: IPv4-to-IPv4 and IPv6-to-IPv6 access between pods
+- Access to External Servers: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to external servers
+- NGINX Ingress Controller Access: Access from IPv4 and/or IPv6 external clients to Kubernetes services via the Kubernetes NGINX Ingress Controller.
```

Comment:
Could you shed some light on why the nginx ingress controller is part of the goals of this KEP? Does that mean there is a commitment to implement the necessary changes (if any) to the nginx ingress controller? Why would we not leave that up to individual ingress controllers to implement?

Reply:
Explained in L456